Consider node shutdown in DataTierAllocationDecider

AthenaEryma opened the pull request:
Prior to this commit, while DataTierAllocationDecider accounted for
Desired Nodes (if available), it did not account for Node Shutdown
information.

As of this commit, nodes that are marked as shutting down for
removal from the cluster are ignored for the purposes of computing
the available data tiers. Desired Nodes configuration, if present,
takes precedence over node shutdown, as desired nodes should be
preferred if possible.

Fixes https://github.com/elastic/elasticsearch/issues/97207

AthenaEryma committed:
Consider node shutdown in DataTierAllocationDecider

Prior to this commit, while DataTierAllocationDecider accounted for
Desired Nodes (if available), it did not account for Node Shutdown
information.

As of this commit, nodes that are marked as shutting down for
removal from the cluster are ignored for the purposes of computing
the available data tiers. Desired Nodes configuration, if present,
takes precedence over node shutdown, as desired nodes should be
preferred if possible.

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/data-tier-decider

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @gwbrown, I've created a changelog YAML for you.

AthenaEryma committed:
Update docs/changelog/98824.yaml

AthenaEryma committed:
Fix changelog

gwbrown commented:
Upon looking at this further, I've removed the `7.17.3` label as the code has diverged significantly enough that a "backport" would really just be an entirely new PR.

idegtiarenko commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -136,14 +148,19 @@ private static Decision debugYesAllowed(RoutingAllocation allocation, List<Strin
 
-        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, nodes);
+        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, filterRemovingNodes(nodes, shutdownMetadata));
     }
```
I am surprised `DataTierAllocationDecider#allocationAllowed(java.lang.String, org.elasticsearch.cluster.node.DiscoveryNode)` is not checking for the node shutdowns, is it already covered by some other decider?

fcofdez replied:
It should be covered by `NodeShutdownAllocationDecider` as you pointed out.

gwbrown replied:
Agreed, that case is covered by `NodeShutdownAllocationDecider`. We tried to keep the logic as isolated as possible, but in this case this decider needs to know about node shutdown so that it won't refuse to allocate shards with a tier preference to nodes without that tier when all the nodes in that tier are shutting down.

idegtiarenko commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -244,4 +261,27 @@ public static boolean allocationAllowed(String tierName, Set<DiscoveryNodeRole>
+        for (var node : removingNodeIds) {
+            builder.remove(node);
+        }
+        return builder.build();
```
Considering https://github.com/elastic/elasticsearch/pull/88436 I wonder if we could avoid having to copy the DiscoveryNodes when calling this allocator with node shutdown in place?

Possibly replace org.elasticsearch.cluster.node.DiscoveryNodes#availableRoles with Map of role -> set of nodeIds and then check (only if there are shutdown entries) that a nodeId exists in the set without corresponding shutdown entry.

gwbrown replied:
Ah, that's a good idea, thanks!

original-brownbear commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:
```diff
@@ -244,4 +261,27 @@ public static boolean allocationAllowed(String tierName, Set<DiscoveryNodeRole>
+     * @return the nodes in the cluster which are not currently marked for removal
+     */
+    private static DiscoveryNodes filterRemovingNodes(DiscoveryNodes nodes, NodesShutdownMetadata nodesShutdownMetadata) {
+        List<String> removingNodeIds = nodesShutdownMetadata.getAll()
```
Looks like this is mostly going to be empty, can we avoid allocating all the stream stuff on the hot path here please? Also, ++ to Ievgen, we shouldn't have to build a new `DiscoveryNodes` here.

original-brownbear reviewed (commented):
Lets try to be a little more performance conscious here please :)

AthenaEryma committed:
Rework per review feedback

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/data-tier-decider

gwbrown commented:
Thanks for the feedback; I'm out of the habit of coding on the hot path like this. I especially appreciate the recommendations for what to do instead - I've removed the `DataNodes` construction in favor of a cached map of roles to nodes in `DiscoveryNodes` and made sure to keep in mind the things mentioned, especially avoiding allocations and avoiding work that isn't strictly necessary.

fcofdez commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -66,7 +67,7 @@ public class DiscoveryNodes implements Iterable<DiscoveryNode>, SimpleDiffable<D
     private final IndexVersion minSupportedIndexVersion;
 
-    private final Set<String> availableRoles;
+    private final Map<String, Set<String>> rolesToNodes;
```
nit: maybe name it `rolesToNodeIds`?

fcofdez commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java:
```diff
@@ -483,13 +597,101 @@ public void testPreferredTierAvailable() {
         }
     }
 
+    public void testDataTierDeciderConsidersNodeShutdown() {
```
can we add a test that goes through the branch where we check that there're still more available nodes in the tier than the number of nodes leaving the cluster?

fcofdez reviewed (commented):
Looks good. I left a few small comments and a suggestion for an extra test.

gwbrown commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -876,14 +876,23 @@ public DiscoveryNodes build() {
+            for (var entry : rolesToNodes.entrySet()) {
+                entry.setValue(Collections.unmodifiableSet(entry.getValue()));
+            }
+            return rolesToNodes;
```
Good catch, thanks - I meant to do that and got distracted by the need to wrap the sets...

AthenaEryma committed:
Address review feedback

Add some more unit tests to make sure all the branches are covered.
Tweak naming.
Make collection immutable

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/data-tier-decider

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -876,14 +876,23 @@ public DiscoveryNodes build() {
+            Map<String, Set<String>> rolesToNodes = new HashMap<>();
+            for (var node : dataNodes.values()) {
+                for (var role : node.getRoles()) {
+                    rolesToNodes.computeIfAbsent(role.roleName(), (key) -> new HashSet<>()).add(node.getId());
```
nit: I think we only care about roles whose `canContainData()` returns `true`. Or really the roles whose names are in `org.elasticsearch.cluster.routing.allocation.DataTier#ALL_DATA_TIERS` plus the legacy `data` role.

gwbrown replied:
The map that's used as the argument to this function is already filtered by `conContainData` and I didn't want to duplicate the check.
https://github.com/elastic/elasticsearch/blob/291ecc5e79f2cf79bb9e2ee1abe9fc7e1b70c923/server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java#L865

DaveCTurner replied:
That filters the _nodes_ but we end up with a map keyed by _role_ many of which are not relevant. For instance if you have a node which has roles `data_hot`, `ingest` and `master` then we'll get all three entries in the resulting map, even though we don't care about the `ingest` and `master` roles.

gwbrown replied:
Ahh, gotcha, you're right. I'll change this.

henningandersen commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java:
```diff
@@ -257,21 +347,37 @@ public void testPreferredTierAvailable() {
             final DesiredNodes desiredNodes = randomBoolean()
                 ? null
                 : createDesiredNodesWithPendingNodes(HOT_DESIRED_NODE, WARM_DESIRED_NODE, COLD_DESIRED_NODE);
+            final var shutdownMetadata = new NodesShutdownMetadata(Collections.emptyMap());
```
Can we make this randomly contain some of the desired nodes too to ensure it does not influence the decision? Also sometimes using irrelevantNodesShutdownMetadata would be good.

gwbrown replied:
👍🏼 I made this test generate much more broad node shutdown metadata, including the existing irrelevant node shutdown metadata as well as (when desired nodes are present) picking node IDs out of the cluster and desired node external IDs to have as marked for removal to make sure those don't influence the decision.

henningandersen reviewed (approved):
LGTM.

fcofdez reviewed (approved):
LGTM

AthenaEryma committed:
More involved shutdown metadata generation in tests, per review

AthenaEryma committed:
Restrict precomuted roles map to data tiers, per review

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into shutdown/data-tier-decider
